crazyflie_sim/crazyflie_sim/backend/np.py

```diff
@@ -0,0 +1,125 @@
+from __future__ import annotations
+
+from rclpy.node import Node
+from rosgraph_msgs.msg import Clock
+from rclpy.time import Time
+from ..simtypes import State, Action
+
+
+import numpy as np
+import rowan
+
+class Backend:
+    """Backend that uses newton-euler rigid-body dynamics implemented in numpy"""
+
+    def __init__(self, node: Node, names: list[str], states: list[State]):
+        self.node = node
+        self.names = names
+        self.clock_publisher = node.create_publisher(Clock, 'clock', 10)
+        self.t = 0
+        self.dt = 0.0005
+
+        self.uavs = []
+        for state in states:
+            uav = Quadrotor(state)
+            self.uavs.append(uav)
+
+    def time(self) -> float:
+        return self.t
+
+    def step(self, states_desired: list[State], actions: list[Action]) -> list[State]:
+        # advance the time
+        self.t += self.dt
+
+        next_states = []
+
+        for uav, action in zip(self.uavs, actions):
+            uav.step(action, self.dt)
+            next_states.append(uav.state)
+
+        # print(states_desired, actions, next_states)
+        # publish the current clock
+        clock_message = Clock()
+        clock_message.clock = Time(seconds=self.time()).to_msg()
+        self.clock_publisher.publish(clock_message)
+
+        return next_states
+
+    def shutdown(self):
+        pass
+
+
+class Quadrotor:
+    """Basic rigid body quadrotor model (no drag) using numpy and rowan"""
+
+    def __init__(self, state):
+        # parameters (Crazyflie 2.0 quadrotor)
+        self.mass = 0.034 # kg
+        # self.J = np.array([
+        # 	[16.56,0.83,0.71],
+        # 	[0.83,16.66,1.8],
+        # 	[0.72,1.8,29.26]
+        # 	]) * 1e-6  # kg m^2
+        self.J = np.array([16.571710e-6, 16.655602e-6, 29.261652e-6])
+
+        # Note: we assume here that our control is forces
+        arm_length = 0.046 # m
+        arm = 0.707106781 * arm_length
+        t2t = 0.006 # thrust-to-torque ratio
+        self.B0 = np.array([
+            [1, 1, 1, 1],
+            [-arm, -arm, arm, arm],
+            [-arm, arm, arm, -arm],
+            [-t2t, t2t, -t2t, t2t]
+            ])
+        self.g = 9.81 # not signed
+
+        if self.J.shape == (3,3):
+            self.inv_J = np.linalg.pinv(self.J) # full matrix -> pseudo inverse
+        else:
+            self.inv_J = 1 / self.J # diagonal matrix -> division
+
+        self.state = state
+
+    def step(self, action, dt):
+
+        # convert RPM -> Force
+        def rpm_to_force(rpm):
+            # polyfit using Tobias' data
```

Comment:
We have them in https://github.com/IMRCLab/crazyflie-system-id (a private repo). I am happy to make it public and refer to it, if the data itself is fine to share publicly?

Reply:
I don't see why not actually! I'll ask tobias just to be sure but since the plots are already out. Perhaps you can already go ahead and refer to it in the text and make the repo public once I received the okay.